File: skills/cp-solver/references/implementations/segment-tree/segment_tree.py
"""
Segment Tree — Generic (Point Update, Range Query)
Supports: sum, min, max, GCD, XOR — any associative operation
Complexity: O(N) build, O(log N) update/query

Usage:
    st = SegmentTree(n)                           # range sum
    st = SegmentTree(n, min, float('inf'))        # range min
    st = SegmentTree(n, max, float('-inf'))       # range max
    st.build(a)         # build from list (0-indexed)
    st.update(i, val)   # set a[i] = val
    st.query(l, r)      # query [l, r] (0-indexed, inclusive)
"""
import sys
input = sys.stdin.readline


class SegmentTree:
    def __init__(self, n, op=lambda a, b: a + b, identity=0):
        self.n = n
        self.op = op
        self.identity = identity
        self.tree = [identity] * (4 * n)

    def build(self, a, node=1, l=0, r=None):
        if r is None: r = self.n - 1
        if l == r:
            self.tree[node] = a[l]
            return
        mid = (l + r) // 2
        self.build(a, 2*node, l, mid)
        self.build(a, 2*node+1, mid+1, r)
        self.tree[node] = self.op(self.tree[2*node], self.tree[2*node+1])

    def update(self, idx, val, node=1, l=0, r=None):
        if r is None: r = self.n - 1
        if l == r:
            self.tree[node] = val
            return
        mid = (l + r) // 2
        if idx <= mid:
            self.update(idx, val, 2*node, l, mid)
        else:
            self.update(idx, val, 2*node+1, mid+1, r)
        self.tree[node] = self.op(self.tree[2*node], self.tree[2*node+1])

    def query(self, ql, qr, node=1, l=0, r=None):
        if r is None: r = self.n - 1
        if qr < l or r < ql:
            return self.identity
        if ql <= l and r <= qr:
            return self.tree[node]
        mid = (l + r) // 2
        return self.op(
            self.query(ql, qr, 2*node, l, mid),
            self.query(ql, qr, 2*node+1, mid+1, r)
        )


# ===== USAGE EXAMPLE =====
def solve():
    import sys
    sys.setrecursionlimit(400000)
    input = sys.stdin.readline

    n, q = map(int, input().split())
    a = list(map(int, input().split()))

    st = SegmentTree(n)          # default: range sum
    st.build(a)

    out = []
    for _ in range(q):
        t, l, r = map(int, input().split())
        if t == 1:
            st.update(l - 1, r)   # 1-indexed → 0-indexed
        else:
            out.append(st.query(l - 1, r - 1))

    sys.stdout.write('\n'.join(map(str, out)) + '\n')


if __name__ == '__main__':
    solve()

File: skills/cp-solver/references/implementations/segment-tree/segment_tree.cpp
// Segment Tree — Generic (Point Update, Range Query)
// Supports: sum, min, max, GCD, XOR — any associative operation
// Complexity: O(N) build, O(log N) update/query
// Usage: Thay op và identity tương ứng với bài toán

#include <bits/stdc++.h>
using namespace std;

template<typename T>
struct SegmentTree {
    int n;
    vector<T> tree;
    T identity;
    function<T(T,T)> op;

    SegmentTree(int n, function<T(T,T)> op, T identity)
        : n(n), op(op), identity(identity) {
        tree.assign(4 * n, identity);
    }

    // Build from array — O(N)
    void build(int node, int l, int r, vector<T>& a) {
        if (l == r) { tree[node] = a[l]; return; }
        int mid = (l + r) / 2;
        build(2*node, l, mid, a);
        build(2*node+1, mid+1, r, a);
        tree[node] = op(tree[2*node], tree[2*node+1]);
    }

    void _update(int node, int l, int r, int idx, T val) {
        if (l == r) { tree[node] = val; return; }
        int mid = (l + r) / 2;
        if (idx <= mid) _update(2*node, l, mid, idx, val);
        else            _update(2*node+1, mid+1, r, idx, val);
        tree[node] = op(tree[2*node], tree[2*node+1]);
    }

    T _query(int node, int l, int r, int ql, int qr) {
        if (qr < l || r < ql) return identity;
        if (ql <= l && r <= qr) return tree[node];
        int mid = (l + r) / 2;
        return op(_query(2*node, l, mid, ql, qr),
                  _query(2*node+1, mid+1, r, ql, qr));
    }

    void build(vector<T>& a)    { build(1, 0, n-1, a); }
    void update(int idx, T val) { _update(1, 0, n-1, idx, val); }
    T    query(int l, int r)    { return _query(1, 0, n-1, l, r); }
};

// ===== USAGE EXAMPLES =====

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, q;
    cin >> n >> q;
    vector<long long> a(n);
    for (auto& x : a) cin >> x;

    // Range Sum
    SegmentTree<long long> st_sum(n, [](long long a, long long b){ return a+b; }, 0LL);

    // Range Min
    SegmentTree<long long> st_min(n, [](long long a, long long b){ return min(a,b); }, LLONG_MAX);

    // Range Max
    SegmentTree<long long> st_max(n, [](long long a, long long b){ return max(a,b); }, LLONG_MIN);

    st_sum.build(a);
    st_min.build(a);
    st_max.build(a);

    while (q--) {
        int type, l, r;
        cin >> type >> l >> r;
        if (type == 1) {
            // Point update: a[l] = r
            st_sum.update(l, r);
        } else {
            // Range query [l, r]
            cout << st_sum.query(l, r) << "\n";
        }
    }
    return 0;
}
